Add support for lzma files, both normal files and tar-files compressed by lzma.
Fixes bug #502999.

 ChangeLog              |   11 +++++++++++
 src/fr-archive.c       |   18 ++++++++++++++++++
 src/fr-command-cfile.c |   24 ++++++++++++++++++++++++
 src/fr-command-tar.c   |   49 +++++++++++++++++++++++++++++++++++++++++++++++++
 src/main.c             |    4 ++++
 src/typedefs.h         |    3 +++
 6 files changed, 109 insertions(+)

Index: src/fr-archive.c
===================================================================
--- src/fr-archive.c	(revision 2059)
+++ src/fr-archive.c	(working copy)
@@ -674,6 +674,14 @@
 		return (archive->command != NULL);
 	}
 
+	if (file_extension_is (filename, ".tar.lzma")
+	    || file_extension_is (filename, ".tzma")) {
+		archive->command = fr_command_tar_new (archive->process,
+						       filename,
+						       FR_COMPRESS_PROGRAM_LZMA);
+		return (archive->command != NULL);
+	}
+
 	if (file_extension_is (filename, ".tar.lzo")
 	    || file_extension_is (filename, ".tzo")) {
 		archive->command = fr_command_tar_new (archive->process,
@@ -763,6 +771,12 @@
 			return (archive->command != NULL);
 		}
 
+		if (file_extension_is (filename, ".lzma")) {
+			archive->command = fr_command_cfile_new (archive->process, filename, FR_COMPRESS_PROGRAM_LZMA);
+			archive->is_compressed_file = TRUE;
+			return (archive->command != NULL);
+		}
+
 		if (file_extension_is (filename, ".lzo")) {
 			archive->command = fr_command_cfile_new (archive->process, filename, FR_COMPRESS_PROGRAM_LZOP);
 			archive->is_compressed_file = TRUE;
@@ -2595,6 +2609,7 @@
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_BZ)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_BZ2)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_GZ)
+		|| (archive->command->file_type == FR_FILE_TYPE_TAR_LZMA)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_LZOP)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_COMPRESS));
 }
@@ -2936,6 +2951,7 @@
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_BZ)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_BZ2)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_GZ)
+		|| (archive->command->file_type == FR_FILE_TYPE_TAR_LZMA)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_LZOP)
 		|| (archive->command->file_type == FR_FILE_TYPE_TAR_COMPRESS));
 }
@@ -3385,6 +3401,7 @@
 		".iso",
 		".jar",
 		".lzh",
+		".lzma",
 		".lzo",
 		".rar",
 		".rpm",
@@ -3393,6 +3410,7 @@
 		".tar.bz",
 		".tar.bz2",
 		".tar.gz",
+		".tar.lzma",
 		".tar.lzo",
 		".tar.Z",
 		".taz",
Index: src/fr-command-cfile.c
===================================================================
--- src/fr-command-cfile.c	(revision 2059)
+++ src/fr-command-cfile.c	(working copy)
@@ -277,6 +277,14 @@
 		fr_process_end_command (comm->process);
 		break;
 
+	case FR_COMPRESS_PROGRAM_LZMA:
+		fr_process_begin_command (comm->process, "lzma");
+		fr_process_set_working_dir (comm->process, temp_dir);
+		fr_process_add_arg (comm->process, "--");
+		fr_process_add_arg (comm->process, filename);
+		fr_process_end_command (comm->process);
+		break;
+
 	case FR_COMPRESS_PROGRAM_LZOP: /* FIXME: untested. */
 		fr_process_begin_command (comm->process, "lzop");
 		fr_process_set_working_dir (comm->process, temp_dir);
@@ -288,6 +296,7 @@
 		break;
 	}
 
+
       	/* copy compressed file to the dest dir */
 
 	fr_process_begin_command (comm->process, "cp");
@@ -398,6 +407,14 @@
 		fr_process_end_command (comm->process);
 		break;
 
+	case FR_COMPRESS_PROGRAM_LZMA:
+		fr_process_begin_command (comm->process, "lzma");
+		fr_process_add_arg (comm->process, "-f");
+		fr_process_add_arg (comm->process, "-d");
+		fr_process_add_arg (comm->process, e_temp_file);
+		fr_process_end_command (comm->process);
+		break;
+
 	case FR_COMPRESS_PROGRAM_LZOP:
 		fr_process_begin_command (comm->process, "lzop");
 		fr_process_set_working_dir (comm->process, temp_dir);
@@ -552,6 +569,11 @@
 		return NULL;
 	}
 
+	if ((prog == FR_COMPRESS_PROGRAM_LZMA) &&
+	    (!is_program_in_path("lzma"))) {
+		return NULL;
+	}
+
 	if ((prog == FR_COMPRESS_PROGRAM_LZOP) &&
 	    (!is_program_in_path("lzop"))) {
 		return NULL;
@@ -569,6 +591,8 @@
 		comm->file_type = FR_FILE_TYPE_BZIP2;
 	else if (prog == FR_COMPRESS_PROGRAM_COMPRESS)
 		comm->file_type = FR_FILE_TYPE_COMPRESS;
+	else if (prog == FR_COMPRESS_PROGRAM_LZMA)
+		comm->file_type = FR_FILE_TYPE_LZMA;
 	else if (prog == FR_COMPRESS_PROGRAM_LZOP)
 		comm->file_type = FR_FILE_TYPE_LZOP;
 
Index: src/fr-command-tar.c
===================================================================
--- src/fr-command-tar.c	(revision 2059)
+++ src/fr-command-tar.c	(working copy)
@@ -212,6 +212,11 @@
 		fr_process_add_arg (comm->process, "-Z");
 		break;
 
+	case FR_COMPRESS_PROGRAM_LZMA:
+		fr_process_add_arg (comm->process,
+				    "--use-compress-program=lzma");
+		break;
+
 	case FR_COMPRESS_PROGRAM_LZOP:
 		fr_process_add_arg (comm->process,
 				    "--use-compress-program=lzop");
@@ -527,6 +532,27 @@
 		new_name = g_strconcat (c_tar->uncomp_filename, ".Z", NULL);
 		break;
 
+	case FR_COMPRESS_PROGRAM_LZMA:
+		fr_process_begin_command (comm->process, "lzma");
+		fr_process_set_sticky (comm->process, TRUE);
+		fr_process_set_begin_func (comm->process, begin_func__recompress, comm);
+		switch (compression) {
+		case FR_COMPRESSION_VERY_FAST:
+			fr_process_add_arg (comm->process, "-1"); break;
+		case FR_COMPRESSION_FAST:
+			fr_process_add_arg (comm->process, "-3"); break;
+		case FR_COMPRESSION_NORMAL:
+			fr_process_add_arg (comm->process, "-6"); break;
+		case FR_COMPRESSION_MAXIMUM:
+			fr_process_add_arg (comm->process, "-9"); break;
+		}
+		fr_process_add_arg (comm->process, "-f");
+		fr_process_add_arg (comm->process, c_tar->uncomp_filename);
+		fr_process_end_command (comm->process);
+
+		new_name = g_strconcat (c_tar->uncomp_filename, ".lzma", NULL);
+		break;
+
 	case FR_COMPRESS_PROGRAM_LZOP:
 		fr_process_begin_command (comm->process, "lzop");
 		fr_process_set_sticky (comm->process, TRUE);
@@ -639,6 +665,13 @@
 			new_name[l - 2] = 0;
 		break;
 
+	case FR_COMPRESS_PROGRAM_LZMA:
+		/* X.tar.lzma --> X.tar 
+		 * (There doesn't seem to be a shorthand suffix) */
+		if (file_extension_is (e_filename, ".tar.lzma"))
+			new_name[l - 5] = 0;
+		break;
+
 	case FR_COMPRESS_PROGRAM_LZOP:
 		/* X.tzo     -->  X.tar
 		 * X.tar.lzo -->  X.tar */
@@ -757,6 +790,17 @@
 		}
 		break;
 
+	case FR_COMPRESS_PROGRAM_LZMA:
+		if (archive_exists) {
+			fr_process_begin_command (comm->process, "lzma");
+			fr_process_set_begin_func (comm->process, begin_func__uncompress, comm);
+			fr_process_add_arg (comm->process, "-f");
+			fr_process_add_arg (comm->process, "-d");
+			fr_process_add_arg (comm->process, tmp_name);
+			fr_process_end_command (comm->process);
+		}
+		break;
+
 	case FR_COMPRESS_PROGRAM_LZOP:
 		if (archive_exists) {
 			fr_process_begin_command (comm->process, "lzop");
@@ -935,6 +979,11 @@
                 return NULL;
         }
 
+        if ((prog == FR_COMPRESS_PROGRAM_LZMA) &&
+            (!is_program_in_path("lzma"))) {
+                return NULL;
+        }
+
         if ((prog == FR_COMPRESS_PROGRAM_LZOP) &&
             (!is_program_in_path("lzop"))) {
                 return NULL;
Index: src/main.c
===================================================================
--- src/main.c	(revision 2059)
+++ src/main.c	(working copy)
@@ -77,6 +77,7 @@
 	{ FR_FILE_TYPE_GZIP,         ".gz",      "application/x-gzip", NULL},
 	{ FR_FILE_TYPE_JAR,          ".jar",     "application/x-jar", N_("Jar (.jar)")},
 	{ FR_FILE_TYPE_LHA,          ".lzh",     "application/x-lha", N_("Lha (.lzh)") },
+	{ FR_FILE_TYPE_LZMA,         ".lzma",    "application/x-lzma", NULL },
 	{ FR_FILE_TYPE_LZOP,         ".lzo",     "application/x-lzop", NULL },
 	{ FR_FILE_TYPE_RAR,          ".rar",     "application/x-rar", N_("Rar (.rar)") },
 	{ FR_FILE_TYPE_RPM,          ".rpm",     "application/x-rpm", NULL },
@@ -84,6 +85,7 @@
 	{ FR_FILE_TYPE_TAR_BZ,       ".tar.bz",  "application/x-bzip-compressed-tar", N_("Tar compressed with bzip (.tar.bz)") },
 	{ FR_FILE_TYPE_TAR_BZ2,      ".tar.bz2", "application/x-bzip-compressed-tar", N_("Tar compressed with bzip2 (.tar.bz2)") },
 	{ FR_FILE_TYPE_TAR_GZ,       ".tar.gz",  "application/x-compressed-tar", N_("Tar compressed with gzip (.tar.gz)") },
+	{ FR_FILE_TYPE_TAR_LZMA,     ".tar.lzma","application/x-lzma-compressed-tar", N_("Tar compressed with lzma (.tar.lzma)") },
 	{ FR_FILE_TYPE_TAR_LZOP,     ".tar.lzo", "application/x-lzop-compressed-tar", N_("Tar compressed with lzop (.tar.lzo)") },
 	{ FR_FILE_TYPE_TAR_COMPRESS, ".tar.Z",   "application/x-compressed-tar", N_("Tar compressed with compress (.tar.Z)") },
 	{ FR_FILE_TYPE_STUFFIT,      ".sit",     "application/x-stuffit", NULL },
@@ -118,6 +120,7 @@
 	{ "zip",        TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_JAR },
 	{ "zip",        TRUE,  FALSE,  TRUE,  FR_FILE_TYPE_EXE },
 	{ "lha",        TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_LHA },
+	{ "lzma",       TRUE,  TRUE,  FALSE, FR_FILE_TYPE_LZMA },
 	{ "lzop",       TRUE,  TRUE,  FALSE, FR_FILE_TYPE_LZOP },
 	{ "rpm2cpio",   TRUE,  FALSE, TRUE,  FR_FILE_TYPE_RPM },
 	{ "uncompress", TRUE,  FALSE, FALSE, FR_FILE_TYPE_COMPRESS },
@@ -132,6 +135,7 @@
 	{ "gzip",      TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_TAR_GZ },
 	{ "bzip2",     TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_TAR_BZ2 },
 	{ "bzip",      FALSE, TRUE,  TRUE,  FR_FILE_TYPE_TAR_BZ },
+	{ "lzma",      TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_TAR_LZMA },
 	{ "lzop",      TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_TAR_LZOP },
 	{ "compress",  TRUE,  TRUE,  TRUE,  FR_FILE_TYPE_TAR_COMPRESS }
 };
Index: src/typedefs.h
===================================================================
--- src/typedefs.h	(revision 2059)
+++ src/typedefs.h	(working copy)
@@ -51,6 +51,7 @@
 	FR_COMPRESS_PROGRAM_BZIP,
 	FR_COMPRESS_PROGRAM_BZIP2,
 	FR_COMPRESS_PROGRAM_COMPRESS,
+	FR_COMPRESS_PROGRAM_LZMA,
 	FR_COMPRESS_PROGRAM_LZOP
 } FRCompressProgram;
 
@@ -93,6 +94,7 @@
 	FR_FILE_TYPE_GZIP,
 	FR_FILE_TYPE_JAR,
 	FR_FILE_TYPE_LHA,
+	FR_FILE_TYPE_LZMA,
 	FR_FILE_TYPE_LZOP,
 	FR_FILE_TYPE_RAR,
 	FR_FILE_TYPE_RPM,
@@ -100,6 +102,7 @@
 	FR_FILE_TYPE_TAR_BZ,
 	FR_FILE_TYPE_TAR_BZ2,
 	FR_FILE_TYPE_TAR_GZ,
+	FR_FILE_TYPE_TAR_LZMA,
 	FR_FILE_TYPE_TAR_LZOP,
 	FR_FILE_TYPE_TAR_COMPRESS,
 	FR_FILE_TYPE_STUFFIT,
